src/docs/arch-concurrency.dox

```diff
@@ -0,0 +1,107 @@
+/*! @arch_page arch-concurrency WiredTiger concurrency management
+
+WiredTiger is a multi-threaded application and needs to manage concurrent access to shared memory
+locations. Many mechanisms are defined in WiredTiger to safely access shared memory locations. This
+page describes what those mechanisms are and the intended usage of them in the code base.
+
+@section volatile Volatile
+
+A controversial tool in the C toolkit. WiredTiger utilizes volatile for one reason, managing
+compiler optimization. Leveraging the definitions from "Is Parallel Programming Hard, And, If So,
+What Can You Do About It?", Paul E. McKenney, WiredTiger utilizes volatile to prevent the
+following type of compiler optimizations: Load fusing, store fusing, invented loads, invented
+stores and code reordering. WiredTiger defines the following macros WT_READ_ONCE and
+WT_WRITE_ONCE, these macros, if used correctly, prevent the described forms of compiler
+optimization. Additionally some variables in WiredTiger are defined with the volatile keyword if
+it is known that they will be frequently accessed in a concurrent context.
+
+@section locking Locking
+
+There are three types of locks in WiredTiger:
+ - Condition Variables
+ - Spin Locks
+ - Read Write (RW) Locks
+Each mechanism guarantees that once a thread relinquishes its lock and another thread obtains the
+same lock, all modifications made by the first thread will be visible to the thread now holding
+the lock.
+
+There are two implementations of condition variables in WiredTiger respectively for POSIX and MSVC
+under the same interface. The specific implementation is determined at compile time based on the
+platform.
+
+There are three different implementations for spin locks in WiredTiger under the same interface:
+ - An implementation using GCC's builtin atomic operations.
+ - An implementation using a POSIX based mutex.
+ - An implementation for MSVC.
+Again specific implementation is determined at compile time based on the platform and compile-time
+configurations. WiredTiger has its own read-write lock implementation. It is based on atomic
+operations and condition variables.
+
+@subsection lock-usage Lock usage
+A correctly used lock takes the lock for each access to the shared memory it guards. WiredTiger
+sometimes deviates from this model and reads variables outside the context of the relevant lock.
+Such instances are only permitted if reading inconsistent data is sufficient for the use case.
+
+@section generations Generations
```

Comment:
Talking about the generation management design is valuable, but I find it incongruous with the other low-level descriptions in this document. Perhaps a later section or separate document with other data tracking mechanisms would flow better.

Reply:
As we do not currently have that document / page, would it be acceptable to move it at a later time? We will create an arch-usage patterns page but that talks about higher level data stuctures and algorithms within WiredTiger, which generations likely doesn't fall into either.

As this page has the heading "WiredTiger concurrency management" I think generations belongs here.

In theory the generation system is interesting enough and could have enough detail to have it's own page in the docs. However writing that page is not within the scope of PM-3221 and would fall under a separate documentation ticket.

Reply:
👍 Fair call.